### PR TITLE
[Fix] Unbuffered streams causing poor performance with MSVC 2015

### PR DESCRIPTION
**Pull request by C-53**
When compiled using Visual Studio 2015 bit7z will use the default buffer size instead of being unbuffered.

## Description

Uses `#if defined(_MSC_VER) && _MSC_VER != 1900` so that CFileOutStream & CFileInStream will use the default buffer size rather than unbuffered file streams for MSVC 2015.

## Motivation and Context

CFileOutStream & CFileInStream were recently modified to call pubsetbuf() with a zero buffer size to improve performance; however, MSVC 2015 appears to have an issue with this. Extracting a [7MB archive](https://github.com/user-attachments/files/17235704/test.zip) on a fast machine took 18 seconds. Following my change the same archive took 0.2 seconds.

I did the same test in Visual Studio 2022 and performance was ~0.2 seconds, so I have not tested reverting the change and have restricted my change to MSVC 2015. I do not know how many MSVC versions between 2015 and 2022 have this problem.

## How Has This Been Tested?

I was unable to build the project tests, but I was able to reproduce the issue in a console application with the following code:

```
void extract(std::string filepath, std::string outpath)
{
	auto start = std::chrono::steady_clock::now();
	bit7z::Bit7zLibrary lib{ "7z.dll" };
	bit7z::BitArchiveReader archive{ lib, filepath, bit7z::BitFormat::Zip };
	archive.extractTo(outpath);
	std::cout << std::chrono::duration_cast<std::chrono::duration<double>>(std::chrono::steady_clock::now() - start).count();
}
```

I used static runtime, Visual Studio 2015 update 3 or Visual Studio 2022 (17.11.4), running on a Windows 11 Pro machine with a 7z.dll from a 7-zip installation.

The underlying issue with streams in 2015 can be observed with std::ifstream in a console application and calling pubsetbuf() with a zero buffer size, and reading files of various sizes.

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by rikyoz** [comment]
Hi!
Yeah, I replicated the issue in MSVC 2015. 
It seems to be a bug in this specific version of MSVC, as I also tested the library with MSVC 2017, and it doesn't have the same issue.
The PR is mostly ok, there's just a small change needed (see the review comment).
I'll merge the PR after the change.
Thank you for the PR!

**Comment by C-53** [comment]
Hi,

I can't see the review comment. Is the review still a draft?

**Review by rikyoz — CHANGES_REQUESTED** [review]

**Comment by rikyoz** [comment]
> Hi,
> 
> I can't see the review comment. Is the review still a draft?

Sorry, I forgot to submit the review.

File: src/internal/cfileinstream.cpp [hunk 1]
```diff
@@ -22,7 +22,11 @@ CFileInStream::CFileInStream( const fs::path& filePath ) : CStdInStream( mFileSt
      * Note: we need to do this before and after opening the file (https://stackoverflow.com/a/59161297/3497024). */
     mFileStream.rdbuf()->pubsetbuf( nullptr, 0 );
     openFile( filePath );
+
+// Unbuffered streams are slow for Visual Studio 2015
+#if defined(_MSVC_VER) && _MSVC_VER != 1900
```
[the remark this answers is not in the document]
Comment (C-53) [reply src/internal/cfileinstream.cpp]: The embarrassing thing is that I'd looked at the documentation and had managed to read it as _MSVC_VER.

File: src/internal/cfileinstream.cpp [hunk 2]
```diff
@@ -22,7 +22,10 @@ CFileInStream::CFileInStream( const fs::path& filePath ) : CStdInStream( mFileSt
      * Note: we need to do this before and after opening the file (https://stackoverflow.com/a/59161297/3497024). */
     mFileStream.rdbuf()->pubsetbuf( nullptr, 0 );
     openFile( filePath );
+// Unbuffered streams are slow for Visual Studio 2015
+#if defined(_MSC_VER) && _MSC_VER != 1900
```
Comment (rikyoz) [inline src/internal/cfileinstream.cpp:26]: Sorry, I just noticed that also the overall condition is not exactly correct. Bit7z can be built also using other compilers, so the condition should be `!defined(_MSC_VER) || _MSC_VER != 1900`. In this way, either the compiler is not MSVC, or if it is MSVC, it is not MSVC 2015. Basically, the code would work as before except for MSVC 2015.

File: src/internal/cfileinstream.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rikyoz) [reply src/internal/cfileinstream.cpp]: No problem :)

**Review by rikyoz — APPROVED** [review]

**Comment by rikyoz** [comment]
Just merged the PR. Thank you again! 🙏

**Pull request merged**